service_container.rst

```diff
@@ -815,20 +815,35 @@ Public Versus Private Services
 From Symfony 4.0, every service defined is private by default.
 
 What does this mean? When a service **is** public, you can access it directly
-from the container object, which is accessible from any controller that extends
-``Controller``::
+from the container object, which can also be injected thanks to autowiring.
+This is mostly useful when you want to fetch services lazily::
 
-    use App\Service\MessageGenerator;
+    namespace App\Service;
 
-    // ...
-    public function new()
+    use Psr\Container\ContainerInterface;
+
+    class MessageGenerator
     {
-        // there IS a public "logger" service in the container
-        $logger = $this->container->get('logger');
+        private $container;
 
-        // this will NOT work: MessageGenerator is a private service
-        $generator = $this->container->get(MessageGenerator::class);
-    }
+        public function __construct(ContainerInterface $container)
+        {
+            $this->container = $container;
+        }
+
+        public function generate(string $message, string $template = null, array $context = []): string
+        {
+            if ($template && $this->container->has('twig')) {
+                // there IS a public "twig" service in the container
+                $twig = $this->container->get('twig');
+
+                return $twig->render($template, $context + ['message' => $message]);
```

Comment:
The + looks a bit strange here. Is it valid php code? If yes, didn’t see it before

Reply:
Yes array union :), see https://www.php.net/manual/en/language.operators.array.php

Reply:
TIL ✌🏻